server/repository/src/migrations/mod.rs

```diff
@@ -0,0 +1,156 @@
+    MigrationAboveAppVersion(Version, Version),
+    #[error("Error during migration ({version})")]
+    MigrationError {
+        source: anyhow::Error,
+        version: Version,
+    },
+    #[error(transparent)]
+    DatabaseError(#[from] RepositoryError),
+}
+
+// TODO logging
+pub fn migrate(
+    connection: &StorageConnection,
+    to_version: Option<Version>,
+) -> Result<Version, MigrationError> {
+    let migrations: Vec<Box<dyn Migration>> = vec![
+        Box::new(V1_00_04),
+        #[cfg(test)]
+        Box::new(templates::adding_table::V1_00_05),
+        #[cfg(test)]
+        Box::new(templates::data_migration::V1_00_06),
+        #[cfg(test)]
+        Box::new(templates::data_and_schema::V1_00_07),
+        #[cfg(test)]
+        Box::new(templates::add_data_from_sync_buffer::V1_00_08),
+    ];
+
+    // Historic diesel migrations
+    run_db_migrations(connection).unwrap();
+
+    // Rust migrations
+    let to_version = to_version.unwrap_or(Version::from_package_json());
+
+    let database_version = get_database_version(connection);
+
+    // for `>` see PartialOrd implementation of Version
+    if database_version > to_version {
+        return Err(MigrationError::DatabaseVersionAboveAppVersion(
+            database_version,
+            to_version,
+        ));
+    }
+
+    if database_version.is_pre_release() {
+        return Err(MigrationError::DatabaseVersionIsPreRelease(
+            database_version,
+        ));
+    }
+
+    for migration in migrations {
+        let migration_version = migration.version();
+
+        if migration_version > to_version {
+            // During test, when we specify to_version manually, we want migrations
+            // to stop at that version, even if there are migrations after specified to_version
+            if cfg!(test) {
+                break;
+            }
+
+            return Err(MigrationError::MigrationAboveAppVersion(
+                migration_version,
+                to_version,
+            ));
+        }
+
+        let database_version = get_database_version(connection);
+
+        // TODO transaction ?
+
+        if migration_version > database_version {
+            log::info!("Running database migration {}", migration_version);
+            migration
+                .migrate(connection)
+                .map_err(|source| MigrationError::MigrationError {
+                    source,
+                    version: migration_version.clone(),
+                })?;
+            set_database_version(connection, &migration_version)?;
+        }
+    }
+
+    set_database_version(connection, &to_version)?;
+    Ok(to_version)
+}
+
+fn get_database_version(connection: &StorageConnection) -> Version {
+    match KeyValueStoreRepository::new(connection).get_string(KeyValueType::DatabaseVersion) {
+        Ok(Some(version_str)) => Version::from_str(&version_str),
+        // Rust migrations start at "1.0.3"
```

Comment:
Before this DatabaseVersion key value type doesn't exist

Reply:
shouldn't this be `1.1.0`? The key was introduced in this PR

Reply:
Even though 1.1.0 would be the final migration for the new migration work, for actual upgrade to use KeyValueType::DatabaseVersion I used 1.0.3, the main reason for this was to allow the example migrations to be testable runnable (which means some versions between changing to use `DatabaseVersion` and 1.1.0), perhaps a comment helps to clarify this ? dc48431c695b11b5d3d5af21295cdb1545f58bf2